pkg/controller/render/machine_config_to_ignition.go

```diff
@@ -0,0 +1,99 @@
+	"github.com/openshift/machine-config-operator/test/helpers"
+
+	"gopkg.in/yaml.v3"
+)
+
+const MCDContentPath = "/etc/machine-config-daemon/mcd_content.json"
+
+func MachineConfigToIgnition(mcSpec *mcSpecv1.MachineConfigSpec) (ign3types.Config, error) {
+	// Config
+	ignition, err := ctrlcommon.ParseAndConvertConfig(mcSpec.Config.Raw)
+	if err != nil {
+		return ign3types.Config{}, fmt.Errorf("parsing Ignition config failed: %w", err)
+	}
+
+	var originFile OriginFile
+	// Extensions
+	supportedExtensions := ctrlcommon.GetSupportedExtensions()
+	for _, mcSpecExtension := range mcSpec.Extensions {
+		originFile.Packages = append(originFile.Packages, supportedExtensions[mcSpecExtension]...)
+	}
+
+	// KernelType
+	if ctrlcommon.CanonicalizeKernelType(mcSpec.KernelType) == ctrlcommon.KernelTypeRealtime {
+		originFile.Packages = append(originFile.Packages, "kernel-rt", "kernel-rt-modules", "kernel-rt-modules-extra")
+		originFile.OverrideRemove = append(originFile.OverrideRemove, "kernel", "kernel-modules", "kernel-modules-extra")
+	}
+
+	// add content from Extensions and KernelType to ignition
+	if !originFile.isEmpty() {
+		ignitionOriginFile, err := originFile.toIgnFile(false)
+		if err != nil {
+			return ign3types.Config{}, fmt.Errorf("could not convert OriginFile to Ignition file: %w", err)
+		}
+		ignition.Storage.Files = append(ignition.Storage.Files, ignitionOriginFile)
+	}
+
+	// KernelArguments
+	// FIPS
+	mcdContent, err := json.Marshal(mcSpec.MCDContent)
+	if err != nil {
+		return ign3types.Config{}, fmt.Errorf("couldn't marshal MCD content: %w", err)
+	}
+	ignition.Storage.Files = append(ignition.Storage.Files, helpers.NewIgnFile(MCDContentPath, string(mcdContent)))
+
+	return ignition, nil
+}
+
+type OriginFile struct {
+	Packages       []string `yaml:"packages"`
+	OverrideRemove []string `yaml:"override-remove"`
+}
+
+func (originFile OriginFile) isEmpty() bool {
+	return len(originFile.Packages) == 0 && len(originFile.OverrideRemove) == 0
+}
+
+func (originFile OriginFile) toIgnFile(allowCompression bool) (ign3types.File, error) {
+	treeFileContents, err := yaml.Marshal(originFile)
+	if err != nil {
+		return ign3types.File{}, fmt.Errorf("failed to marshal extensions as yaml: %w", err)
+	}
+	fullYamlContents := append([]byte("# Generated by the MCO\n\n"), treeFileContents...)
+	src, gzipped, err := baseutil.MakeDataURL(fullYamlContents, nil, allowCompression)
+	if err != nil {
+		return ign3types.File{}, fmt.Errorf("could not encode extensions file: %w", err)
+	}
+	hash := sha256.New()
+	hash.Write([]byte(src))
+	sha := hex.EncodeToString(hash.Sum(nil))[0:7]
+	file := ign3types.File{
+		Node: ign3types.Node{
+			Path: "/etc/rpm-ostree/origin.d/extensions-" + sha + ".yaml",
+		},
+		FileEmbedded1: ign3types.FileEmbedded1{
+			Contents: ign3types.Resource{
+				Source: util.StrToPtr(src),
+			},
+			Mode: util.IntToPtr(0644),
+		},
+	}
+	if gzipped {
+		file.Contents.Compression = util.StrToPtr("gzip")
+	}
+
+	return file, nil
+}
```

Comment:
@jmarrero @jlebon this is mostly just copy pasted from @jmarrero 's code, but I did tweak it to allow OverrideRemove.
1. Is there somewhere we could factor this out to? Seems like "create an Ignition file with the first 8 characters of the hash of its contents added to its path" is not a very clean API. Maybe I just need to get over that for now :joy: 
2. Is this correct syntax for override-remove?

Reply:
3. I'm using github.com/coreos/butane/base/util and github.com/coreos/ignition/v2/config/util, is that okay? On a related note, I wonder if some Ignition helpers should be moved to a lib? Eg the MCO has helpers like https://github.com/openshift/machine-config-operator/blob/master/test/helpers/helpers.go#L155. That helper uses EncodeBytes from 	github.com/vincent-petithory/dataurl while the Butane code is using baseutil.MakeDataURL. Which is leading to different encodings being used, which I ran into when writing the unit tests

Reply:
1. The reason I used a SHA with the first characters is trying to mimic how github shows commits. The SHA is not really important the important part is that the filename is unique and reproducible. If the MCO is doing something similar somewhere else or have a better approach it's ok to change it. rpm-ostree does not care about the filename. I am not sure if there is a good place to put shared code or if this is big enough to put in a library of sorts.

Reply:
Re. 2, `override-remove` is [not supported yet in the CoreOS layering flow](https://github.com/coreos/rpm-ostree/issues/3364), but I'm working on it! :)

Reply:
Also https://github.com/coreos/rpm-ostree/issues/2542

Reply:
Having a chat with @mkenigs about this, one idea here is to update the MCO's vendor of fcct :arrow_right: butane, and then generate butane for this fragment.  @mkenigs is suggesting alternatively that butane exposes a Go API to generate the origin file directly.

Reply:
Chatted with @mkenigs about this too. Generating a Butane fragment and then rendering that seems fine, but personally my vote is to keep those ~30 duplicated lines here for now as we iterate on getting something working and add a `TODO` we can circle back to. Once we gain more confidence in how everything is connected, we can start looking at optimizing and deduping things.